src/utils/JSONReviver.ts

```diff
@@ -1,85 +1,100 @@
 /* Generic Reviver, toJSON, and fromJSON functions used for saving and loading objects */
 
-import { validateObject } from "./Validator";
+import { ObjectValidator, validateObject } from "./Validator";
 
 export interface IReviverValue {
   ctor: string;
   data: any;
 }
 
-// A generic "smart reviver" function.
-// Looks for object values with a `ctor` property and
-// a `data` property. If it finds them, and finds a matching
-// constructor that has a `fromJSON` property on it, it hands
-// off to that `fromJSON` function, passing in the value.
-export function Reviver(key: string, value: IReviverValue | null): any {
-  if (value == null) {
-    return null;
-  }
+/**
+ * A generic "smart reviver" function.
+ * Looks for object values with a `ctor` property and a `data` property.
+ * If it finds them, and finds a matching constructor, it hands
+ * off to that `fromJSON` function, passing in the value. */
+export function Reviver(key: string, value: unknown): any {
+  if (typeof value !== "object" || value === null) return value;
+  // @ts-ignore We know it's an object here, just checking type of object properties to type validate.
+  if (typeof value.ctor !== "string" || typeof value.data === "undefined") return value;
+  const typedValue = value as IReviverValue;
 
-  if (typeof value === "object" && typeof value.ctor === "string" && typeof value.data !== "undefined") {
-    // Compatibility for version v0.43.1
-    // TODO Remove this eventually
-    if (value.ctor === "AllServersMap") {
-      console.warn("Converting AllServersMap for v0.43.1");
-      return value.data;
+  const ctor = Reviver.constructors[typedValue.ctor];
+  if (!ctor) {
+    // Known missing constructors with special handling.
+    switch (typedValue.ctor) {
+      case "AllServersMap":
+        console.warn("Converting AllServersMap for v0.43.1");
+        return typedValue.data;
     }
+    // Missing constructor with no special handling. Throw error.
+    throw new Error(`Could not locate constructor named ${typedValue.ctor}. If the save data is valid, this is a bug.`);
+  }
 
-    const ctor = Reviver.constructors[value.ctor];
-
-    if (typeof ctor === "function" && typeof ctor.fromJSON === "function") {
-      const obj = ctor.fromJSON(value);
-      if (ctor.validationData !== undefined) {
-        validateObject(obj, ctor.validationData);
-      }
-      return obj;
-    }
+  const obj = ctor.fromJSON(typedValue);
+  if (ctor.validationData !== undefined) {
+    validateObject(obj, ctor.validationData);
   }
-  return value;
+  return obj;
 }
 
 // eslint-disable-next-line @typescript-eslint/no-namespace
 export namespace Reviver {
```

Comment:
Yeah I was actually going to do this, I only didn't because it would need to edit a bunch of files that are ultimately unrelated. There's no reason the constructors need to be a property of the Revivers function

Reply:
Done in 34b5b77